libs/constrained-generators/testlib/Test/Min/Base.hs

```diff
@@ -0,0 +1,501 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- Base types: Term, Pred, Spec, Ctx, and classes: HasSpec, Syntax, Semantics, and Logic for the MinModel
+module Test.Min.Base where
+
+import Constrained.Core (Evidence (..), Var (..), eqVar)
+import Constrained.GenT
+import Constrained.List hiding (ListCtx)
+import Data.Kind
+import Data.List.NonEmpty (NonEmpty (..))
+import qualified Data.List.NonEmpty as NE
+import Data.String (fromString)
+import Data.Typeable
+import GHC.Stack
+import Prettyprinter
+
+-- ===========================================
+-- Terms
+-- ===========================================
+
+type AppRequires t dom rng =
+  ( Logic t
+  , TypeList dom
+  , Eq (t dom rng)
+  , Show (t dom rng)
+  , Typeable dom
+  , Typeable rng
+  , All HasSpec dom
+  , HasSpec rng
+  )
+
+data Term a where
+  App ::
+    forall t dom rng.
+    AppRequires t dom rng =>
+    t dom rng ->
+    List Term dom ->
+    Term rng
+  Lit :: (Typeable a, Eq a, Show a) => a -> Term a
+  V :: HasSpec a => Var a -> Term a
+
+instance Eq (Term a) where
+  V x == V x' = x == x'
+  Lit a == Lit b = a == b
+  App (w1 :: x1) (ts :: List Term dom1) == App (w2 :: x2) (ts' :: List Term dom2) =
+    case (eqT @dom1 @dom2, eqT @x1 @x2) of
+      (Just Refl, Just Refl) ->
+        w1 == w2
+          && sameTerms ts ts'
+      _ -> False
```

Comment:
Another thing I'd really like to do here is just have `App1` and `App2` for unary and binary application so that we can get rid of `List` for this implementation. That would in turn simplify some of this code (specifically `sameTerms`) furthermore doing the "only one function universe" approach would allow us to get rid of `eqT @x1 @x2` here.